pkg/diagnose/runner_test.go

```diff
@@ -8,22 +8,71 @@ package diagnose
 import (
 	"bytes"
 	"errors"
+	"regexp"
 	"testing"
 
 	"github.com/DataDog/datadog-agent/pkg/diagnose/diagnosis"
 
 	"github.com/stretchr/testify/assert"
 )
 
-func TestRunAll(t *testing.T) {
+func TestConnectivityAutodiscovery(t *testing.T) {
 
-	diagnosis.Register("failing", func() error { return errors.New("fail") })
-	diagnosis.Register("succeeding", func() error { return nil })
+	diagnosis.RegisterMetadataAvail("failing", func() error { return errors.New("fail") })
+	diagnosis.RegisterMetadataAvail("succeeding", func() error { return nil })
 
 	w := &bytes.Buffer{}
-	RunAll(w)
+	RunMetadataAvail(w)
 
 	result := w.String()
 	assert.Contains(t, result, "=== Running failing diagnosis ===\n===> FAIL")
 	assert.Contains(t, result, "=== Running succeeding diagnosis ===\n===> PASS")
 }
+
+func TestDiagnoseAllBasicRegAndRunNoDiagnoses(t *testing.T) {
+
+	diagnosis.Register("TestDiagnoseAllBasicRegAndRunNoDiagnoses", func(cfg diagnosis.DiagnoseConfig) []diagnosis.Diagnosis {
+		return nil
+	})
+
+	re, _ := regexp.Compile("TestDiagnoseAllBasicRegAndRunNoDiagnoses")
+	diagCfg := diagnosis.DiagnoseConfig{
+		Include: []*regexp.Regexp{re},
+	}
+	diagnoses := RunAll(diagCfg)
+	assert.Len(t, diagnoses, 0)
+}
+
+func TestDiagnoseAllBasicRegAndRunSomeDiagnosis(t *testing.T) {
+
+	inDiagnoses := []diagnosis.Diagnosis{
+		{
+			Result:      diagnosis.DiagnosisSuccess,
+			Name:        "Name_foo",
+			Category:    "Category_foo",
+			Description: "Description_foo",
+			Remediation: "Remediation_foo",
+			RawError:    errors.New("Error_foo"),
+		},
+		{
+			Result:      diagnosis.DiagnosisFail,
+			Name:        "Name_bar",
+			Category:    "Category_bar",
+			Description: "Description_bar",
+			Remediation: "Remediation_bar",
+			RawError:    errors.New("Error_bar"),
+		},
+	}
+
+	diagnosis.Register("TestDiagnoseAllBasicRegAndRunSingleDiagnosis", func(cfg diagnosis.DiagnoseConfig) []diagnosis.Diagnosis {
+		return inDiagnoses
+	})
+
+	re, _ := regexp.Compile("TestDiagnoseAllBasicRegAndRunSingleDiagnosis")
+	diagCfg := diagnosis.DiagnoseConfig{
+		Include: []*regexp.Regexp{re},
+	}
+	outSuitesDiagnoses := RunAll(diagCfg)
+	assert.Len(t, outSuitesDiagnoses, 1)
+	assert.Equal(t, outSuitesDiagnoses[0].SuiteDiagnoses, inDiagnoses)
+}
```

Comment:
Yes. I did. And further improve include/exclude algorithm to have both applicable (per comment above)

Reply:
> Thank you for this 👍
> 
> Have a couple of nits/questions: diangosisCounters -> diagnosisCounters (unless diangosis is a word I don't know)
> 
> Also have some questions:
> 
> The diagnose implementation in pkg/diagnose/check.go recreates the components needed to do a check (aggregator, forwarder, ...) but will in the future directly get the diagnose from the running agent service ? The current version correspond more to the local/cli diagnose of a check ?
> Is there a plan to include [HTTP tracing](https://pkg.go.dev/net/http/httptrace) ([code](https://github.com/DataDog/datadog-agent/blob/3527a4381ecbb1ce26dfd00949f2ccda2bea577d/pkg/diagnose/connectivity/hooks.go#L32)) in the diagnose() function later ? I think it can provide useful information and some remediation steps in some specific cases.
> 
> 
Yes, exactly. In the future I will get diagnose from running agent. This is the reason for --local command line option. I just implemented it first because it was easier and because I need to unblock @alopezz.

Yes for HTTP tracing. One of the reason I "removed" it from the implementation because I had a problem to suppress intermediate output. But it has lot of value indeed. I am adding this a the JIRA for backlog (see current backlog here https://datadoghq.atlassian.net/jira/software/projects/ASUP/boards/4915/backlog